plugins/kapt3/kapt3-base/src/org/jetbrains/kotlin/kapt3/base/annotationProcessing.kt

```diff
@@ -40,6 +41,13 @@ fun KaptContext.doAnnotationProcessing(
 
     val compilerAfterAP: JavaCompiler
     try {
+        if (javaSourceFiles.isEmpty() && aggregatedTypes.isEmpty() && additionalSources.isEmpty()) {
```

Comment:
`aggregatedTypes.isEmpty()` a bit strange for me, do it make sense on classpath changes in case when aggregation processor presented? (Then could we test such case?)

Reply:
Or if same case is covered by `additionalSources.isEmpty` check could first one be removed?

Reply:
`additionalSources` is only used in tests, so I need all 3 checks to be present.

I'll rename `aggregatedTypes` to `typesToProcess` as `doAnnotationProcessing` should not be aware of isolating/aggregating APs. It should just know that there are some types to process.

It is possible for `aggregatedTypes` to contain some elements, while the sources are empty. This is in case that isolating AP generated `AGen.class`, which contains aggregating annotation that generated `AGen2.java`, and in `AGen2.java` a classpath type `X` is mentioned. If in the second run `X` changes, we need to reprocess `AGen.class`.

That being said, I'll add a test case to cover this.

Reply:
It's still not clear why it should be reprocessed by APs/kapt, as I understand any change in X shouldn't affect annotation processor, but just AGen2.java recompilation

Reply:
or it could be in case when AP itself tries to inspect X internals?

Reply:
Case in which `X` internals are inspected is precisely why `T` is reprocessed. E.g for an AP under incremental rules, it is possible to get type `T` which is annotated with a claimed annotation, and then to explore types visible from `T` and to make decisions based on that. The only limitation is that we assume if `T` uses `X` and `X` uses `Y`, `Y` needs to be in ABI of `X` in order to impact `T`. This is how Gradle handles it, and we are doing the same in kapt.